pkg/provisioner/conditionwatcher.go

```diff
@@ -0,0 +1,170 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package provisioner
+
+import (
+	"context"
+	"os"
+	"sync"
+	"time"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2"
+
+	"k8s.io/client-go/tools/cache"
+	azDiskClientSet "sigs.k8s.io/azuredisk-csi-driver/pkg/apis/client/clientset/versioned"
+	azurediskInformers "sigs.k8s.io/azuredisk-csi-driver/pkg/apis/client/informers/externalversions"
+)
+
+const (
+	azvolumeattachment = "azvolumeattachments"
+	azvolume           = "azvolume"
+	azdrivernode       = "azdrivernode"
+)
+
+type waitResult struct {
+	obj runtime.Object
+	err error
+}
+
+type waitEntry struct {
+	conditionFunc func(obj interface{}) (bool, error)
+	waitChan      chan waitResult
+}
+
+type conditionWatcher struct {
+	informerFactory azurediskInformers.SharedInformerFactory
+	waitMap         sync.Map // maps namespaced name to waitEntry
+	namespace       string
+}
+
+func newConditionWatcher(ctx context.Context, azDiskClient azDiskClientSet.Interface, namespace string, resync time.Duration) *conditionWatcher {
+	informerFactory := azurediskInformers.NewSharedInformerFactory(azDiskClient, resync)
+	azVolumeAttachmentInformer := informerFactory.Disk().V1alpha1().AzVolumeAttachments().Informer()
+	azVolumeInformer := informerFactory.Disk().V1alpha1().AzVolumes().Informer()
+	azDriverNodeInformer := informerFactory.Disk().V1alpha1().AzDriverNodes().Informer()
+
+	c := conditionWatcher{
+		informerFactory: informerFactory,
+		waitMap:         sync.Map{},
+	}
+
+	azVolumeAttachmentInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		UpdateFunc: c.onUpdate,
+		DeleteFunc: c.onDelete,
```

Comment:
Shouldn't we also be handling create events?

Reply:
create event doesn't really affect `crdProvisioner` so I removed it from the equation to reduce the total number of event handling calls. Should I add it in case we need create event for future changes to the driver?

Reply:
Calls to the server or just internal ones? I do not believe handling creates would result in additional calls to the server. My understanding is that the watch events are multiplexed.

Reply:
Just internal ones

Reply:
I don't believe that to be a significant concern. It would be better for completeness to include create events even if ultimately they do not satisfy current condition checks.